### PR TITLE
Add manifest entrypoint and database tables

### PR DESCRIPTION
**Pull request by hemanthnakkina**
Add support for manifests in sunbeam.

Manifest option in bootstrap command

```sunbeam cluster bootstrap --accept-defaults --manifest /home/ubuntu/manifest.yaml```

Manifest option in refresh command

```sunbeam cluster refresh --manifest /home/ubuntu/manifest.yaml```

To clear the existing manifest use the following option
 
```sunbeam cluster refresh --clear-manifest```

Currently only specifying local terraform plans are supported as part of manifest file.
```
terraform:
  deploy-openstack:
    source: /home/ubuntu/sunbeam-terraform
```

### CODE REVIEW

File: sunbeam-python/sunbeam/jobs/manifest.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+    @classmethod
+    def load(cls, manifest_file: Path) -> "Manifest":
+        try:
+            with manifest_file.open() as file:
+                return Manifest(**yaml.safe_load(file))
+        except FileNotFoundError as e:
+            raise e
+        except ValidationError as e:
+            raise e
```
Comment (gboutry) [inline sunbeam-python/sunbeam/jobs/manifest.py:76]: There's no value in just re-raising exceptions, you could either get rid of the whole try, or do something about exceptions

File: sunbeam-microcluster/api/manifests.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+	manifests, err := sunbeam.ListManifests(s)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	return response.SyncResponse(true, manifests)
+}
+
+func cmdManifestsGet(s *state.State, r *http.Request) response.Response {
```
Comment (gboutry) [inline sunbeam-microcluster/api/manifests.go:45]:
```suggestion
func cmdManifestGet(s *state.State, r *http.Request) response.Response {
```

File: sunbeam-microcluster/api/manifests.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+	err = sunbeam.AddManifest(s, req.ManifestID, req.Data)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	return response.EmptySyncResponse
+}
+
+func cmdManifestsDelete(s *state.State, r *http.Request) response.Response {
```
Comment (gboutry) [inline sunbeam-microcluster/api/manifests.go:80]:
```suggestion
func cmdManifestDelete(s *state.State, r *http.Request) response.Response {
```

File: sunbeam-python/sunbeam/jobs/manifest.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    @classmethod
+    def load_latest_from_cluserdb(cls) -> "Manifest":
```
[the remark this answers is not in the document]
Comment (hemanthnakkina) [reply sunbeam-python/sunbeam/jobs/manifest.py]: Done in next commit

File: sunbeam-python/sunbeam/jobs/manifest.py (hunk 1 above)
Comment (hemanthnakkina) [reply sunbeam-python/sunbeam/jobs/manifest.py]: Removed exceptions

File: sunbeam-microcluster/api/manifests.go (hunk 3 above)
Comment (hemanthnakkina) [reply sunbeam-microcluster/api/manifests.go]: Modified as per the suggestion

**Comment by hemanthnakkina** [comment]
> Minor nits, otherwise looks good
> 
> Are we sure we want to add pydantic as a dependency ? (I'm all for it)

Pydantic supports nested dataclasses along with the validation. However in python dataclasses the nested attributes are returned as dictionary and not validated to nested classes as well.
Currently pydantic 1.10 is used but subsequent versions provide certain functions like field_validators and model_validators which will be useful in future.

File: sunbeam-microcluster/api/manifests.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/openstack-snaps/snap-openstack/sunbeam-microcluster/api/types"
+	"github.com/openstack-snaps/snap-openstack/sunbeam-microcluster/sunbeam"
+)
+
+// /1.0/manifests endpoint.
+var manifestsCmd = rest.Endpoint{
+	Path: "manifests",
+
+	Get:  rest.EndpointAction{Handler: cmdManifestsGetAll, ProxyTarget: true},
+	Post: rest.EndpointAction{Handler: cmdManifestPost, ProxyTarget: true},
```
Comment (gboutry) [inline sunbeam-microcluster/api/manifests.go:23]:
```suggestion
	Post: rest.EndpointAction{Handler: cmdManifestsPost, ProxyTarget: true},
```
This one should stay plural

File: sunbeam-python/sunbeam/jobs/manifest.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+from pydantic.dataclasses import dataclass
+
+from sunbeam.clusterd.client import Client as clusterClient
+from sunbeam.jobs.common import BaseStep, Result, ResultType, Status
+
+LOG = logging.getLogger(__name__)
+EMPTY_MANIFEST = """charms: {}
+terraform-plans: {}
+"""
+VALID_TERRAFORM_PLANS = [
```
Comment (gboutry) [inline sunbeam-python/sunbeam/jobs/manifest.py:31]: I don't think that's the right way, implemented like this, we have to know all the plans sunbeam is using, on every plugins (even external if we want to be able to override them)

File: sunbeam-microcluster/api/manifests.go (hunk 5 above)
Comment (hemanthnakkina) [reply sunbeam-microcluster/api/manifests.go]: Done

File: sunbeam-python/sunbeam/jobs/manifest.py (hunk 6 above)
Comment (hemanthnakkina) [reply sunbeam-python/sunbeam/jobs/manifest.py]: Modified so that the plugin can specify the keys and used as part of validation.

**Comment by hemanthnakkina** [comment]
Closed in favour of https://github.com/canonical/snap-openstack/pull/73

**Pull request closed without merging**